pkg/specgen/generate/namespaces.go

```diff
@@ -49,51 +48,26 @@ func GetDefaultNamespaceMode(nsType string, cfg *config.Config, pod *libpod.Pod)
 		}
 	}
 
-	// If we have containers.conf and are not using cgroupns, use that.
-	if cfg != nil && nsType != "cgroup" {
-		switch nsType {
-		case "pid":
-			return specgen.ParseNamespace(cfg.Containers.PidNS)
-		case "ipc":
-			return specgen.ParseNamespace(cfg.Containers.IPCNS)
-		case "uts":
-			return specgen.ParseNamespace(cfg.Containers.UTSNS)
-		case "user":
-			return specgen.ParseUserNamespace(cfg.Containers.UserNS)
-		case "net":
-			ns, _, err := specgen.ParseNetworkNamespace(cfg.Containers.NetNS)
-			return ns, err
-		}
+	if cfg == nil {
+		cfg = &config.Config{}
 	}
-
 	switch nsType {
-	case "pid", "ipc", "uts":
-		// PID, IPC, UTS both default to private, do nothing
+	case "pid":
+		return specgen.ParseNamespace(cfg.Containers.PidNS)
+	case "ipc":
+		return specgen.ParseNamespace(cfg.Containers.IPCNS)
+	case "uts":
+		return specgen.ParseNamespace(cfg.Containers.UTSNS)
 	case "user":
-		// User namespace always defaults to host
-		toReturn.NSMode = specgen.Host
-	case "net":
-		// Net defaults to Slirp on rootless, Bridge otherwise.
-		if rootless.IsRootless() {
-			toReturn.NSMode = specgen.Slirp
-		} else {
-			toReturn.NSMode = specgen.Bridge
-		}
+		return specgen.ParseUserNamespace(cfg.Containers.UserNS)
 	case "cgroup":
-		// Cgroup is host for v1, private for v2.
-		// We can't trust c/common for this, as it only assumes private.
-		cgroupsv2, err := cgroups.IsCgroup2UnifiedMode()
-		if err != nil {
-			return toReturn, err
-		}
-		if !cgroupsv2 {
-			toReturn.NSMode = specgen.Host
-		}
-	default:
-		return toReturn, errors.Wrapf(define.ErrInvalidArg, "invalid namespace type %s passed", nsType)
+		return specgen.ParseCgroupNamespace(cfg.Containers.CgroupNS)
+	case "net":
+		ns, _, err := specgen.ParseNetworkNamespace(cfg.Containers.NetNS)
+		return ns, err
```

Comment:
Maybe I'm not following the logic, but don't you need a rootless check here?  Will Slirp still be returned?

Reply:
Containers.conf takes care of setting the default to slirp4netns or bridge already.  So that should not be a problem.  Matt points out above ,that if for some reason we do not have a conf and the default is "", then we need the rootless check in ParseNetworkNamespace, which I fixed.